### PR TITLE
JAMES-XXXX JMAP Filter/get implementation

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    |           },
+                    |           "c1"]
+                    |      ]
+                    |}""".stripMargin
+
+    val response = `given`()
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala:66]: debug statement

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import io.restassured.http.ContentType.JSON
+import net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson
+import org.apache.http.HttpStatus
+import org.apache.james.GuiceJamesServer
+import org.apache.james.jmap.http.UserCredential
+import org.apache.james.jmap.rfc8621.contract.Fixture.{ACCEPT_RFC8621_VERSION_HEADER, BOB, BOB_PASSWORD, DOMAIN, authScheme, baseRequestSpecBuilder}
+import org.apache.james.utils.DataProbeImpl
+import org.junit.jupiter.api.{BeforeEach, Test}
+
+trait FilterGetMethodContract {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala:33]: Can you think of adding more test cases?

Have you looked at tests written for the draft version to get an idea of what had already been done on the topic?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/Capability.scala [hunk 3]
```diff
@@ -43,6 +43,7 @@ object CapabilityIdentifier {
   val EMAIL_SUBMISSION: CapabilityIdentifier = "urn:ietf:params:jmap:submission"
   val JAMES_QUOTA: CapabilityIdentifier = "urn:apache:james:params:jmap:mail:quota"
   val JAMES_SHARES: CapabilityIdentifier = "urn:apache:james:params:jmap:mail:shares"
+  val JMAP_FILTER: CapabilityIdentifier = "urn:apache:james:params:jmap:mail:filter"
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/Capability.scala:46]: JMAP will be a linagora specific extension hence we should use `"urn:linagora:openpaas:params:jmap:mail:filter"`

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/filter/FilterGet.scala [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.james.jmap.method.WithAccountId
+import org.apache.james.mailbox.model.MailboxId
+
+case class FilterGetRequest(accountId: AccountId,
+                            ids: List[Id]) extends WithAccountId
+
+case class FilterGetResponse(accountId: AccountId,
+                             list: List[Filter])
+
+case class Condition(field: String, comparator: String, value: String)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/filter/FilterGet.scala:34]: We should strong type all these fields with case classes

**Review by chibenwa — COMMENTED** [review]
Let's target a feature branch instead of master, while waiting to have the openpaas-james repository setted up...

**Comment by quantranhong1999** [comment]
Have 3 tests green for now. I will write more tests in the next coming days.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala [hunk 5]
```diff
@@ -211,4 +210,91 @@ trait FilterGetMethodContract {
...
+                     |   "methodCalls": [
+                     |       [
+                     |           "Filter/get",
+                     |           {
+                     |                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+                     |                "ids": ["singleton"]
+                     |           },
+                     |           "c1"]
+                     |      ]
+                     |}""".stripMargin
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala:227]: Do we need that much of a big indent?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala [hunk 6]
```diff
@@ -211,4 +210,91 @@ trait FilterGetMethodContract {
...
+         |  "methodResponses": [[
+         |    "error",
+         |      {
+         |        "type": "invalidArguments",
+         |        "description": "{\\"errors\\":[{\\"path\\":\\"obj.accountId\\",\\"messages\\":[\\"error.expected.jsstring\\"]}]}"
+         |      },
+         |    "c1"]]
+         |}""".stripMargin)
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/FilterGetMethodContract.scala:299]: What happens when we use another id than `singleton` ?

**Comment by Arsnael** [comment]
test this please

**Pull request closed without merging**